### ISSUE TITLE
Enable new APIs

### ISSUE THREAD

**Issue by avaranovich** [Azure/acs-engine#178]
To be able to run chron jobs, it is required to have batch/v2alpha1 enabled on the API server
http://kubernetes.io/docs/user-guide/cron-jobs/

would be good to have it enabled on ACS as well, or have a guide how to do enable it manually.

**Comment by josibake** [comment]
+1

**Comment by colemickens** [comment]
I don't think ACS is going to be enabling alpha-level features.

You can edit the templates here in ACS-Engine and easily add this ([in the code, here](https://github.com/Azure/acs-engine/blob/master/parts/kubernetesmastercustomdata.yml#L96)). In the future we can look at adding this into the apimodel as well so that you wouldn't have to make the templates changes manually.

**Comment by avaranovich** [comment]
enabling some king of OOB experimental mode flag would be a good idea to have the latest kubernetes bits on ACS.

**Comment by avaranovich** [comment]
I applied the following change https://gist.github.com/avaranovich/c1f944343b98e7627d0265b85ca641f1#file-kubernetesmastercustomdata-yml-L111
after re-building the necessary artifacts with ./acs-engine examples/kubernetes.classic.json
I deployed the template with Azure CLI 2.0.

Now, if I try to deploy a job
https://gist.github.com/avaranovich/c3a5784d08bb7aca42991e27f48314ab

I am still getting 
`error validating "test-job.yaml": error validating data: couldn't find type: v2alpha1.ScheduledJob; if you choose to ignore these errors, turn validation off with --validate=false `

**Comment by josibake** [comment]
@colemickens I can second what Andrei just described. Changing the template did not appear to change anything.

**Comment by colemickens** [comment]
You have to rebuild the binary when you change template parts.

**Comment by avaranovich** [comment]
After changing the part, I did "go build" to re-build acs-engine, then re-generated the templates. Did I miss some steps? I understand, the content of the parts gets embedded as binary custom data into the templates, but I was not sure what are the excact steps to get custom data updated.

**Comment by colemickens** [comment]
Sorry @avaranovich I meant to reply much earlier.

Yes, you can run `make` or run the regeneration first and the build second, but `make` is easiest.

**Comment by anhowe** [comment]
A generalized feature request here can be to add custom parameters to the apiserver, kubelet, and controller.  Prioritizing as P2, and will watch for interest on this feature.

**Comment by josibake** [comment]
just a note, took me a second to work around this:

had to run `go get -u github.com/jteeuwen/go-bindata/...` before `make` would work

**Comment by colemickens** [comment]
Yeah. Building with `go get .` requires bindata to run. `bindata` requires go-bindata to be installed and in PATH. I have some cleanup around here that should make this easier, so that really running `make` is totally sufficient (I guess assuming that `$GOPATH` is setup and `$GOPATH/bin` in on `$PATH`... but those are fairly standard for Go development.

And of course, using the docker container should be easiest as everything is included and ready to go.

**Comment by JonathanFraser** [comment]
This is a deployment tool, users need to be able to deploy alpha clusters so that they can vet their code against the new features. This is 100% must requirement. While I agree it shouldn't be default, there should be an option when deploying that doesn't require mucking about with code.

**Comment by Karnich** [comment]
@avaranovich if i wanna enable batch/v2alpha1 on a cluster that is already up and running, can i just add it to the kube-apiserver.yaml file on the master node and then restart the VM? or even just edit it from the dasboard?

**Comment by colemickens** [comment]
@Karnich Yes.

@JonathanFraser pull requests accepted. So far, it's really trivial for users to make these changes in a hacky manner to unblock themselves. Plumbing it up in the API as an adjustable flag is considerably less trivial. We're not opposed to it, but we don't have the time right now to chase down enabling experimental features...

**Comment by rossedman** [comment]
@Karnich @avaranovich @colemickens To enable this does it require actually restarting the VM or do we just need to restart the api server containers? Also does this require any changes or flags added to the kubelets?

**Comment by Karnich** [comment]
@rossedman Im not 100% sure but i restarted my VM. Wasnt really a problem to restart the master since its NoSchedule

**Comment by rossedman** [comment]
@Karnich I just dropped the change in the manifests folder and it picked it up and appears to be working.

**Comment by dtzar** [comment]
I've tested and used this.  You can pass `--runtime-config` and other things when you define your initial API model configuration and see the default values passed in [this document](https://github.com/Azure/acs-engine/blob/master/docs/clusterdefinition.md#feat-apiserver-config). 

For instance, this model configuration adds Initilizers support:
```
{
  "apiVersion": "vlabs",
  "properties": {
    "orchestratorProfile": {
      "orchestratorType": "Kubernetes",
      "orchestratorRelease": "1.9",
      "kubernetesConfig": {
        "apiServerConfig" : {
          "--admission-control": "NamespaceLifecycle,LimitRanger,ServiceAccount,DefaultStorageClass,ResourceQuota,DenyEscalatingExec,AlwaysPullImages,SecurityContextDeny,Initializers",
          "--runtime-config": "admissionregistration.k8s.io/v1alpha1"
        }
      }
    },
```

Per @avaranovich it also looks like what you desire is automatically included in k8s 1.8+ so you don't need to override.
```
CronJob resource in batch/v2alpha1 API group has been deprecated starting from cluster version 1.8. You should switch to using batch/v1beta1, instead, which is enabled by default in the API server. 
```

Someone please correct me if I'm wrong, but I believe we can close this issue out.
